Components/Queues/Drivers/PhpAmqpLibDriver.php

```diff
@@ -0,0 +1,308 @@
+
+            $this->connectionsData[] = [
+                'host' => $parsedUrl['host'] ?? self::DEFAULT_HOST,
+                'user' => $username,
+                'password' => $password,
+                'vhost' => $vhost,
+                'port' => $parsedUrl['port'] ?? self::DEFAULT_PORT,
+            ];
+        }
+    }
+
+    /**
+     * Method responsible to open the connection with the broker.
+     *
+     * @param bool $shuffle - used to randomize the connections if exists more than one
+     *
+     * @throws \Exception
+     */
+    public function connect($shuffle = true)
+    {
+        $this->shuffleConnections($shuffle);
+        try {
+            $this->stream = AMQPStreamConnection::create_connection($this->connectionsData, []);
+        } catch (AMQPIOException $exception) {
+            throw new \AMQPConnectionException($exception->getMessage());
+        }
+    }
+
+    /**
+     * Responsible to disconnect with the broker.
+     *
+     * @throws \Exception
+     * @return void
+     */
+    public function disconnect()
+    {
+        if ($this->validateConnection()) {
+            $this->stream->close();
+        }
+    }
+
+    /**
+     * Creates the QueueMessage object.
+     *
+     * @return QueueMessageInterface
+     */
+    public function createQueueMessage(): QueueMessageInterface
+    {
+        $msg = new QueueMessage();
+        $msg->setContext(new Context());
+
+        return $msg;
+    }
+
+    /**
+     * Verifies if there is some connection opened with the broker.
+     *
+     * @return bool
+     */
+    public function isConnected(): bool
+    {
+        return $this->validateConnection() && $this->stream->isConnected();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function destroy(string $consumerTag)
+    {
+        $this->channel->basic_cancel($consumerTag);
+        $this->disconnect();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function consume(string $consumerTag, string $queueName, callable $callback = null)
+    {
+        $this->declareChannel();
+        $this->declareQueue($queueName);
+        $this->channel->basic_consume($queueName, $consumerTag, false, false, false, false, $callback);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function ack(QueueMessageInterface $message = null)
+    {
+        $this->channel->basic_ack($message->getMessageId());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function nack(QueueMessageInterface $message = null)
+    {
+        $this->channel->basic_nack($message->getMessageId());
+    }
+
+    /**
+     * Creates a message and prepare the content to publish.
+     *
+     * {@inheritdoc}
+     * @throws \Exception
+     */
+    public function send(QueueMessageInterface $message, $destination = null, array $arguments = []): bool
+    {
+        $this->declareChannel();
+        $this->declareQueue($message->getQueue(), AMQP_DURABLE, $arguments);
```

Comment:
AMQP_DURABLE is only available if ext-amqp is installed, which we don't require in our composer.json nor we validate it while building the container. Should we define this internally, require ext-amqp?

Reply:
```
➜  integration-framework-bundle git:(php-amqp-lib-v2) docker run php:7.2 php -r "echo AMQP_DURABLE;"

Warning: Use of undefined constant AMQP_DURABLE - assumed 'AMQP_DURABLE' (this will throw an Error in a future version of PHP) in Command line code on line 1
AMQP_DURABLE%  
```

Reply:
If we don't specify "ext-amqp" do we have an alternative in PHP AMQP lib for that?

Reply:
We can define it ourselved. It's just a constant for the number 2. Not sure if the library already defines it. @brunohsouza?

Reply:
If the value is 2 we already have this constant in QueueMessage DELIVERY_MODE_PERSISTENT